### PR TITLE
ScheduledJob tests

### PR DESCRIPTION
**Pull request by soltysh**
This builds on top of #25569, so only the last two commits matter ([Revert commit a31ca0d and move batch/v2alpha1 tests to separate group](https://github.com/kubernetes/kubernetes/commit/e7f6ba796e2aecf8f0560342fd0fae8d0865dc53) and [ScheduledJob client and storage tests](https://github.com/kubernetes/kubernetes/commit/c59c045a8c6081eda2cdd27e7fdbcc481144d3b5)). This also supersedes #25566 with simpler approach to testing, by testing single version at a time. 
@caesarxuchao I've reverted the changes I've did previously to enable multi-version tests and introduced new group where `batch/v2alpha1` is tested (see 1st commit). I've also added back the necessary tests for scheduledjob related stuff (2nd commit) that are already in queue (@erictune fyi - I can't stand having code without tests :wink:).

[Analytics]()

### CODE REVIEW

**Comment by caesarxuchao** [comment]
@soltysh Sorry, I'm OOO this week and have a deadline to meet. Do you mind if I review it next week?

**Comment by soltysh** [comment]
@caesarxuchao although ScheduledJob (which this PR is part of) is targeted for 1.3 as well, I think nothing will happen if the review of the tests from here will be done next week. Besides, I doubt everything I need will be merged this week, the queue is failing quite frequently recently. One other option is that @erictune will pick up the review here, but I'd prefer him to go after ScheduledJob pieces. Anyway, thanks for the heads up :)

**Comment by soltysh** [comment]
@caesarxuchao with 1.3 deadline closing later today will you be able to have a look at this PR? I've rebased it on top of current master so it contains only the required bits.

**Comment by soltysh** [comment]
@k8s-bot unit test this github: #IGNORE

**Comment by soltysh** [comment]
Rebased due to https://github.com/kubernetes/kubernetes/issues/25953

**Comment by soltysh** [comment]
Rebased due to https://github.com/kubernetes/kubernetes/pull/26063.

**Comment by caesarxuchao** [comment]
@soltysh I was OOO until last night. The changes to the test api seems good. Do you want someone else to review the actual tests?

File: pkg/api/testapi/testapi.go [hunk 1]
```diff
@@ -63,10 +63,9 @@ var (
 )
 
 type TestGroup struct {
-	// the first element in the group is meant to be the preferred version
-	externalGroupVersions []unversioned.GroupVersion
```
Comment (caesarxuchao) [inline pkg/api/testapi/testapi.go]: It seems you were undoing your changes, the testapi used to only have one externalGroupVersion, am I right?

**Comment by erictune** [comment]
LGTM

File: pkg/api/testapi/testapi.go (hunk 1 above)
Comment (soltysh) [inline pkg/api/testapi/testapi.go]: Correct.

Comment (caesarxuchao) [inline pkg/api/testapi/testapi.go]: LGTM.

**Comment by soltysh** [comment]
@erictune @caesarxuchao thx!

File: hack/test-go.sh [hunk 2]
```diff
@@ -58,7 +58,7 @@ KUBE_GOVERALLS_BIN=${KUBE_GOVERALLS_BIN:-}
 # Lists of API Versions of each groups that should be tested, groups are
 # separated by comma, lists are separated by semicolon. e.g.,
 # "v1,compute/v1alpha1,experimental/v1alpha2;v1,compute/v2,experimental/v1alpha3"
-KUBE_TEST_API_VERSIONS=${KUBE_TEST_API_VERSIONS:-"v1,extensions/v1beta1,metrics/v1alpha1,federation/v1alpha1;v1,autoscaling/v1,batch/v1,batch/v2alpha1,extensions/v1beta1,apps/v1alpha1,metrics/v1alpha1,federation/v1alpha1,policy/v1alpha1"}
+KUBE_TEST_API_VERSIONS=${KUBE_TEST_API_VERSIONS:-"v1,extensions/v1beta1,metrics/v1alpha1,federation/v1alpha1;v1,autoscaling/v1,batch/v1,extensions/v1beta1,apps/v1alpha1,metrics/v1alpha1,federation/v1alpha1,policy/v1alpha1;v1,autoscaling/v1,batch/v2alpha1,extensions/v1beta1,apps/v1alpha1,metrics/v1alpha1,federation/v1alpha1,policy/v1alpha1"}
```
Comment (lavalamp) [inline hack/test-go.sh]: Do we need to test three ways? I don't think we can make our unit tests take another 20 minutes.

Comment (caesarxuchao) [inline hack/test-go.sh]: If we have https://github.com/kubernetes/kubernetes/pull/25566/files, we may need only one setting. And a unit test can test multiple versions in the go code if it really cares.

Comment (soltysh) [inline hack/test-go.sh]: @lavalamp with the current testapi implementation we can't support multiversions in tests :disappointed: From my investigations there are two options: either we add entire group, which as you mentioned will significantly increase the tests time (and is only temporary solution); or we need to rework testapi which I've started in the PR @caesarxuchao linked, but from what I've seen it will require quite a lot of changes, since testapi is spread all over the code. Personally I'm happy with either approach, although the one from this PR is simpler. Let's have this in as is and we can work on improving testapi to support multiversions, which will be required at some point in time when more groups will support additional versions. Agreed?

**Comment by alex-mohr** [comment]
buildcop says: this seems to make tests better, so marking p2 so it jumps the queue.

File: hack/test-go.sh (hunk 2 above)
Comment (lavalamp) [inline hack/test-go.sh]: I have a stack of PRs mostly done which fixes the registered package-- one
of the next steps was going to be to replace the testapi package with a
function that lets you set up the real registration stuff however you like.
Unfortunately, it's not going to make it for 1.3. So please do whatever you
have to, but:

1. Unit tests can't get slower than they already are
2. Try not to leave me with something that's really hard to untangle later
:)

Comment (caesarxuchao) [inline hack/test-go.sh]: @lavalamp can we omit the first setting? I don't think we have unit tests that are specifically designed to run when many groups are not registered.

**Comment by caesarxuchao** [comment]
@alex-mohr does this one fix unit tests? 

Anyway, this one will add 50% running time of unit tests. See https://github.com/kubernetes/kubernetes/pull/25737#discussion_r64482001. I'm removing the p2 for now.

**Comment by alex-mohr** [comment]
@caesarxuchao Looks like I misread this -- going quickly through 60+ pending PRs to prioritize test-related ones.   Thanks for removing p2

File: hack/test-go.sh (hunk 2 above)
Comment (lavalamp) [inline hack/test-go.sh]: I wanted to test with everything optional turned off. However: maybe? Need
to investigate. I'm fairly skeptical that the testapi package is performing
a useful service at the moment.

Comment (caesarxuchao) [inline hack/test-go.sh]: @soltysh @lavalamp I think we can remove the first setting. Because:
1. the testapi package imports all the installs anyway: https://github.com/kubernetes/kubernetes/blob/master/pkg/api/testapi/testapi.go#L40
2. all optional groups are accessible through testapi anyway:
https://github.com/kubernetes/kubernetes/blob/master/pkg/api/testapi/testapi.go#L134

Comment (lavalamp) [inline hack/test-go.sh]: @mikedanese who is looking at making our unit tests faster

Comment (soltysh) [inline hack/test-go.sh]: I'm inclined to drop this from 1.3 and wait for @lavalamp changes. This PR introduces only tests to ScheduledJobs, which without the controller that didn't make the 1.3 deadline isn't of much use. I agree with Daniel that the tests are already too long and making it longer isn't a desired thing. 

> Try not to leave me with something that's really hard to untangle later

Yes sir :)

> I think we can remove the first setting.

That's viable option for now. I'll try to do it later today.

Comment (caesarxuchao) [inline hack/test-go.sh]: I removed the 1.3 label. Sorry for the trouble.

Comment (soltysh) [inline hack/test-go.sh]: OK, @lavalamp can you ping me once your testapi changes land, please?

Comment (lavalamp) [inline hack/test-go.sh]: I hope to get back to those soon :)

File: pkg/client/unversioned/scheduledjobs_test.go [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+	"k8s.io/kubernetes/pkg/apis/batch"
+	"k8s.io/kubernetes/pkg/apis/batch/v2alpha1"
+	"k8s.io/kubernetes/pkg/client/unversioned/testclient/simple"
+)
+
+func getScheduledJobsResource() string {
+	return "scheduledjobs"
+}
+
+func TestListScheduledJob(t *testing.T) {
```
Comment (caesarxuchao) [inline pkg/client/unversioned/scheduledjobs_test.go]: @soltysh is this test similar to other tests in the folder, which verify that various client methods work? If so, then it might not be necessary, becasue the generated clientset test all the methods. We plan to deprecate the pkg/client/unversioned and ask people to use clientset in the near future. It's just a FYI, we can still have these tests merged.

File: pkg/registry/scheduledjob/etcd/etcd_test.go [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+	"k8s.io/kubernetes/pkg/apis/batch/v2alpha1"
+	"k8s.io/kubernetes/pkg/fields"
+	"k8s.io/kubernetes/pkg/labels"
+	"k8s.io/kubernetes/pkg/registry/generic"
+	"k8s.io/kubernetes/pkg/registry/registrytest"
+	"k8s.io/kubernetes/pkg/runtime"
+	etcdtesting "k8s.io/kubernetes/pkg/storage/etcd/testing"
+)
+
+func newStorage(t *testing.T) (*REST, *StatusREST, *etcdtesting.EtcdTestServer) {
```
Comment (caesarxuchao) [inline pkg/registry/scheduledjob/etcd/etcd_test.go]: @erictune does this test still look good to you? You LGTMed it 2 weeks ago.

Comment (caesarxuchao) [inline pkg/registry/scheduledjob/etcd/etcd_test.go]: Or we can separate the test to another PR.

**Comment by caesarxuchao** [comment]
@soltysh if you copy Mike's change to KUBE_TEST_APIVERSIONS, and remove the two test files (for client and etcd) you added here, I can lgtm it again and apply the 1.3 label. I think the tests don't really have to be v1.3 and I'm a little nervous to merge 500+ lines at this point. Thanks.

**Comment by soltysh** [comment]
@caesarxuchao ptal, I've moved the 2nd commit to #27157

**Comment by caesarxuchao** [comment]
LGTM. Thanks. 

@mikedanese FYI.

**Comment by erictune** [comment]
I agree with earlier comment that we don't need test since we are moving to
generated.

**Comment by jszczepkowski** [comment]
@soltysh 
The PR was reverted by #27184 as it was causing integration test failure.

**Comment by caesarxuchao** [comment]
The integration failure seems should happen consistently, instead of being a flake. @soltysh do you have time to fix it? The [log](https://console.cloud.google.com/storage/browser/kubernetes-jenkins/logs/kubernetes-test-go/12909/) and the failure:
```
=== RUN   TestBatchGroupBackwardCompatibility
--- FAIL: TestBatchGroupBackwardCompatibility (0.04s)
	master_test.go:293: case {POST /apis/batch/v2alpha1/namespaces/default/jobs 
		{
		    "kind": "Job",
		    "apiVersion": "batch/v1",
		    "metadata": {
		        "name": "pi"
		    },
		    "spec": {
		        "parallelism": 1,
		        "completions": 1,
		        "template": {
		            "metadata": {
		                "name": "pi",
		                "creationTimestamp": null
		            },
		            "spec": {
		                "containers": [
		                    {
		                        "name": "pi",
		                        "image": "perl",
		                        "command": [
		                            "perl",
		                            "-Mbignum=bpi",
		                            "-wle",
		                            "print bpi(2000)"
		                        ]
		                    }
		                ],
		                "restartPolicy": "Never"
		            }
		        }
		    }
		}
		 map[201:true] }
	master_test.go:294: Expected status one of map[201:true], but got 404
	master_test.go:295: Body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"the server could not find the requested resource","reason":"NotFound","details":{},"code":404}
	master_test.go:293: case {GET /apis/batch/v2alpha1/namespaces/default/jobs/pi  map[200:true] batch/v2alpha1}
	master_test.go:294: Expected status one of map[200:true], but got 404
	master_test.go:295: Body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"the server could not find the requested resource","reason":"NotFound","details":{},"code":404}
	master_test.go:298: case {GET /apis/batch/v2alpha1/namespaces/default/jobs/pi  map[200:true] batch/v2alpha1}
	master_test.go:299: Expected version batch/v2alpha1, got body {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"the server could not find the requested resource","reason":"NotFound","details":{},"code":404}
	master_test.go:293: case {GET /apis/extensions/v1beta1/namespaces/default/jobs/pi  map[200:true] extensions/v1beta1}
	master_test.go:294: Expected status one of map[200:true], but got 404
	master_test.go:295: Body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"jobs.batch \"pi\" not found","reason":"NotFound","details":{"name":"pi","group":"batch","kind":"jobs"},"code":404}
	master_test.go:298: case {GET /apis/extensions/v1beta1/namespaces/default/jobs/pi  map[200:true] extensions/v1beta1}
	master_test.go:299: Expected version extensions/v1beta1, got body {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"jobs.batch \"pi\" not found","reason":"NotFound","details":{"name":"pi","group":"batch","kind":"jobs"},"code":404}
	master_test.go:293: case {DELETE /apis/batch/v2alpha1/namespaces/default/jobs/pi  map[200:true] v1}
	master_test.go:294: Expected status one of map[200:true], but got 404
	master_test.go:295: Body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"the server could not find the requested resource","reason":"NotFound","details":{},"code":404}
	master_test.go:293: case {GET /apis/batch/v2alpha1/namespaces/default/jobs/pi  map[200:true] batch/v2alpha1}
	master_test.go:294: Expected status one of map[200:true], but got 404
	master_test.go:295: Body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"the server could not find the requested resource","reason":"NotFound","details":{},"code":404}
	master_test.go:298: case {GET /apis/batch/v2alpha1/namespaces/default/jobs/pi  map[200:true] batch/v2alpha1}
	master_test.go:299: Expected version batch/v2alpha1, got body {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"the server could not find the requested resource","reason":"NotFound","details":{},"code":404}
=== RUN   TestAccept
--- PASS: TestAccept (0.04s)
	master_test.go:338: body: kind: APIVersions
		serverAddressByClientCIDRs:
		- clientCIDR: 0.0.0.0/0
		  serverAddress: ""
		versions:
		- v1
	master_test.go:356: body: {"kind":"APIVersions","versions":["v1"],"serverAddressByClientCIDRs":[{"clientCIDR":"0.0.0.0/0","serverAddress":""}]}
```

**Comment by soltysh** [comment]
Not sooner than next week since I'm traveling back to PL over the weekend.

**Comment by soltysh** [comment]
@caesarxuchao do we still want this for 1.3?

**Comment by mikedanese** [comment]
Yes please.

**Comment by caesarxuchao** [comment]
Yes, thank you @soltysh. Let me know if you need help.

**Comment by soltysh** [comment]
Done in the linked PR.

**Comment by caesarxuchao** [comment]
Thank you @soltysh
